### ISSUE TITLE
Implement New Model: Graph Deep Factors for Forecasting

### ISSUE THREAD

**Issue by StatMixedML** [awslabs/gluon-ts#1116]
# Description

The following is taken from [Graph Deep Factors for Forecasting](https://arxiv.org/abs/2010.07373):

> Deep probabilistic forecasting techniques have recently been proposed for modeling large collections of time-series. However, these techniques explicitly assume either complete independence (local model) or complete dependence (global model) between time-series in the collection. This corresponds to the two extreme cases where every time-series is disconnected from every other time-series in the collection or likewise, that every time-series is related to every other time-series resulting in a completely connected graph. In this work, we propose a deep hybrid probabilistic graph-based forecasting framework called Graph Deep Factors (GraphDF) that goes beyond these two extremes by allowing nodes and their time-series to be connected to others in an arbitrary fashion. GraphDF is a hybrid forecasting framework that consists of a relational global and relational local model.  In particular, we propose a relational global model that learns complex non-linear time-series patterns globally using the structure of the graph to improve both forecasting accuracy and computational efficiency. Similarly, instead of modeling every time-series independently, we learn a relational local model that not only considers its individual time-series but also the time-series of nodes that are connected in the graph.

The idea is to have a global-local model that explicitly considers the local pattern of each time series, which is in contrast to purely global models, such as DeepAR, MQRNN, etc. Also, it is a very nice extension of the already available Deep Factor model.

![image](https://user-images.githubusercontent.com/41187941/97445201-4bbf8f00-192d-11eb-96eb-ad5f662de931.png)

**Comment by StatMixedML** [comment]
@lostella Kindly asking if the implementation of the model is in any way a priority of your group? Adding this model would be very beneficial.

Thanks!

**Comment by dcmaddix** [comment]
Yes, this would be a great extension of our Deep Factor model.  Thank you!

**Comment by lostella** [comment]
@StatMixedML no current plan on implementing it, so if you want to give it a go I can assign this to you. Otherwise I'll update you as soon as someone picks this up

**Comment by StatMixedML** [comment]
@lostella I`d love to work on it, but I would first need to get sufficiently familiar with Graph Neural Networks + mxnet. Not sure by when I would be able to start actually implementing the paper. What would be your desired timeplan?

**Comment by timoschowski** [comment]
One comment: this may be much easier to implement in PyTorch -- and that's getting increasingly easy to use with GluonTS.

**Comment by StatMixedML** [comment]
Even though I`d love to, I am really not sure I can commit to a continuous development of the model, since I potentially lack the time to work on it.

**Comment by lostella** [comment]
@StatMixedML I'm keeping this open since it's a feature request, hope that's okay

**Comment by StatMixedML** [comment]
Absolutely, thanks for re-opening!

I wish I could find the time to work on it ...
